src/test/ui/destructuring-assignment/struct-or-enum-variant-path.rs

```diff
@@ -0,0 +1,11 @@
+// check-pass
+
+fn main() {
+    let mut a;
+
+    struct S;
+    (S, a) = (S, ());
+
+    enum E { A }
+    (E::A, a) = (E::A, ());
+}
```

Comment:
Could you add test cases for qualified paths, type-relative paths and `Self` constructor?
```rust
struct S;

impl S {
    fn check() { (Self, a) = (S, ()); }
}

enum E { V }
type A = E;

fn check() {
    (<E>::V, a) = (E::V, ());
    (A::V, a) = (E::V, ());
}
```